File: mobile/lib/features/notifications/data/models/notification_preference.dart
class NotificationPreference {
  final int? id;
  final int? userId;
  final bool emailEnabled;
  final bool pushEnabled;
  final bool smsEnabled;
  final bool websocketEnabled;
  final String? pushEndpoint;
  final String? pushProvider;

  const NotificationPreference({
    this.id,
    this.userId,
    required this.emailEnabled,
    required this.pushEnabled,
    required this.smsEnabled,
    required this.websocketEnabled,
    this.pushEndpoint,
    this.pushProvider,
  });

  factory NotificationPreference.fromJson(Map<String, dynamic> json) {
    return NotificationPreference(
      id: json['id'] as int?,
      userId: json['user_id'] as int?,
      emailEnabled: json['email_enabled'] as bool? ?? true,
      pushEnabled: json['push_enabled'] as bool? ?? true,
      smsEnabled: json['sms_enabled'] as bool? ?? false,
      websocketEnabled: json['websocket_enabled'] as bool? ?? true,
      pushEndpoint: json['push_endpoint'] as String?,
      pushProvider: json['push_provider'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'email_enabled': emailEnabled,
        'push_enabled': pushEnabled,
        'sms_enabled': smsEnabled,
        'websocket_enabled': websocketEnabled,
      };

  NotificationPreference copyWith({
    bool? emailEnabled,
    bool? pushEnabled,
    bool? smsEnabled,
    bool? websocketEnabled,
    String? pushEndpoint,
    String? pushProvider,
  }) {
    return NotificationPreference(
      id: id,
      userId: userId,
      emailEnabled: emailEnabled ?? this.emailEnabled,
      pushEnabled: pushEnabled ?? this.pushEnabled,
      smsEnabled: smsEnabled ?? this.smsEnabled,
      websocketEnabled: websocketEnabled ?? this.websocketEnabled,
      pushEndpoint: pushEndpoint ?? this.pushEndpoint,
      pushProvider: pushProvider ?? this.pushProvider,
    );
  }
}

File: backend/src/apps/notification/schemas/notification_preference.py
"""Pydantic schemas for NotificationPreference."""
from typing import Optional

from pydantic import BaseModel, Field, field_serializer

from src.apps.iam.utils.hashid import encode_id


class NotificationPreferenceRead(BaseModel):
    id: int
    user_id: int
    websocket_enabled: bool
    email_enabled: bool
    push_enabled: bool
    sms_enabled: bool
    push_endpoint: Optional[str] = None
    push_provider: Optional[str] = None
    push_providers: list[str] = Field(default_factory=list)

    model_config = {"from_attributes": True}

    @field_serializer("id", "user_id")
    def serialize_ids(self, value: int) -> str:
        return encode_id(value)


class NotificationPreferenceUpdate(BaseModel):
    """Update channel-level flags. Only provided fields are changed."""

    websocket_enabled: Optional[bool] = None
    email_enabled: Optional[bool] = None
    push_enabled: Optional[bool] = None
    sms_enabled: Optional[bool] = None


class PushSubscriptionUpdate(BaseModel):
    """
    Payload sent by the browser after the user grants push permission.
    Maps to the standard PushSubscription JSON object.
    """

    endpoint: str
    p256dh: str
    auth: str
